src/coreclr/jit/fgprofile.cpp

```diff
@@ -244,387 +465,407 @@ class ClassProbeVisitor final : public GenTreeVisitor<ClassProbeVisitor<TFunctor
+    // Remember the schema index for this block.
     //
-    bool instrument = true;
+    block->bbClassSchemaIndex = (int)schema.size();
 
-    if (!SUCCEEDED(res))
+    // Scan the statements and identify the class probes
+    //
+    BuildClassProbeSchemaGen                    schemaGen(schema, m_schemaCount);
+    ClassProbeVisitor<BuildClassProbeSchemaGen> visitor(m_comp, schemaGen);
+    for (Statement* stmt : block->Statements())
     {
-        JITDUMP("Unable to instrument -- block counter allocation failed: 0x%x\n", res);
-        instrument = false;
-        // The E_NOTIMPL status is returned when we are profiling a generic method from a different assembly
-        if (res != E_NOTIMPL)
-        {
-            noway_assert(!"Error: failed to allocate profileBlockCounts");
-            return;
-        }
+        visitor.WalkTree(stmt->GetRootNodePointer(), nullptr);
+    }
+}
+
+//------------------------------------------------------------------------
+// ClassProbeInstrumentor::Instrument: add class probes to block
+//
+// Arguments:
+//   block -- block of interest
+//   schema -- instrumentation schema
+//   profileMemory -- profile data slab
+//
+void ClassProbeInstrumentor::Instrument(BasicBlock* block, Schema& schema, BYTE* profileMemory)
+{
+    if ((block->bbFlags & BBF_HAS_CLASS_PROFILE) == 0)
+    {
+        return;
     }
 
-    // For each BasicBlock (non-Internal)
-    //  1. Assign the blocks bbCodeOffs to the ILOffset field of this blocks profile data.
-    //  2. Add an operation that increments the ExecutionCount field at the beginning of the block.
+    // Would be nice to avoid having to search here by tracking
+    // candidates more directly.
     //
-    // Each (non-Internal) block has it own BlockCounts tuple [ILOffset, ExecutionCount]
-    // To start we initialize our current one with the first one that we allocated
+    JITDUMP("Scanning for calls to profile in " FMT_BB "\n", block->bbNum);
+
+    // Scan the statements and add class probes
     //
-    int currentSchemaIndex = 0;
+    int classSchemaIndex = block->bbClassSchemaIndex;
+    assert((classSchemaIndex >= 0) && (classSchemaIndex < (int)schema.size()));
 
-    // Hold the address of the first blocks ExecutionCount
-    size_t addrOfFirstExecutionCount = 0;
+    ClassProbeInserter                    insertProbes(schema, profileMemory, &classSchemaIndex, m_instrCount);
+    ClassProbeVisitor<ClassProbeInserter> visitor(m_comp, insertProbes);
+    for (Statement* stmt : block->Statements())
+    {
+        visitor.WalkTree(stmt->GetRootNodePointer(), nullptr);
+    }
+}
 
-    for (block = fgFirstBB; (block != nullptr); block = block->bbNext)
+//------------------------------------------------------------------------
+// ClassProbeInstrumentor::SuppressProbes: clean up if we're not instrumenting
+//
+// Notes:
+//   Currently we're hijacking the gtCallStubAddre of the call node to hold
```

Comment:
nit: => gtStubCallStubAddre => gtStubCallStubAddr

Reply:
Thanks. Will fix in the next PR...